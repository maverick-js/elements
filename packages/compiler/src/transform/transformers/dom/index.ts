import {
  $,
  replaceTsNodes,
  resetArgsCount,
  splitImportsAndBody,
  type TsNodeMap,
} from '@maverick-js/ts';
import ts from 'typescript';

import { Scope } from '../../../parse/ast';
import { markCustomElements } from '../element';
import type { Transformer } from '../transformer';
import { DomTransformState } from './state';
import { transform } from './transform';
import { DomTemplateVariables } from './vars';

export interface DomTransformOptions {
  /**
   * Whether to mark components with a static `tagName` property declaration to be built as
   * a custom element. The custom element itself needs to be registered separately.
   */
  customElements?: boolean;
}

export function domTransformer({ customElements = false }: DomTransformOptions = {}): Transformer {
  return {
    name: '@maverick-js/dom',
    transform({ sourceFile, nodes, ctx }) {
      const hydratable = Boolean(ctx.options.hydratable),
        state = new DomTransformState(null, { hydratable }),
        replace: TsNodeMap = new Map();

      for (const node of nodes) {
        const result = transform(node, state.child(node, new Scope()));
        if (result) replace.set(node.node, result);
        resetArgsCount();
      }

      const { imports, body } = splitImportsAndBody(sourceFile);

      if (ctx.options.delegateEvents && state.delegatedEvents.size > 0) {
        body.push(delegateEvents(state));
      }

      const statements: ts.Statement[] = [];

      const components = ctx.analysis.components;
      for (const component of Object.keys(components)) {
        if (components[component]) state.runtime.add(component);
      }

      const templates = createTemplateVariables(state);

      if (state.runtime.identifiers.length > 0) {
        statements.push(state.runtime.toImportDeclaration());
      }

      if (templates) statements.push(templates.toStatement());

      const transformedSourceFile = replaceTsNodes(
        $.updateSourceFile(sourceFile, [...imports, ...statements, ...state.renders, ...body]),
        replace,
      );

      return customElements ? markCustomElements(transformedSourceFile) : transformedSourceFile;
    },
  };
}

function createTemplateVariables(state: DomTransformState) {
  const templates: [template: ts.Identifier, html: string][] = [],
    vars = new DomTemplateVariables(state.runtime);

  state.walk(({ template, html }) => {
    if (template && html) templates.push([template, html]);
  });

  for (const [template, html] of templates) {
    vars.template(template, html);
  }

  // Dedupe templates.
  for (let i = 0; i < templates.length; i++) {
    const templateA = templates[i];
    for (let j = i + 1; j < templates.length; j++) {
      const templateB = templates[j];
      if (templateA[1] === templateB[1]) {
        vars.update(templateB[0], templateA[0]);
      }
    }
  }

  return templates.length > 0 ? vars : null;
}

function delegateEvents({ delegatedEvents, runtime }: DomTransformState) {
  return $.createExpressionStatement(
    runtime.delegateEvents(
      $.createArrayLiteralExpression(Array.from(delegatedEvents).map((type) => $.string(type))),
    ),
  );
}
