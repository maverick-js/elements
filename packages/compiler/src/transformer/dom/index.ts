import { encode } from 'html-entities';
import kleur from 'kleur';

import { escape } from '../../utils/html';
import {
  createFunctionCall,
  createStringLiteral,
  Declarations,
  selfInvokingFunction,
  trimQuotes,
  trimTrailingSemicolon,
} from '../../utils/print';
import {
  type AST,
  type AttributeNode,
  type ComponentChildren,
  type ElementNode,
  isAST,
  isAttributeNode,
  isAttributesEnd,
  isChildrenEnd,
  isChildrenStart,
  isDirectiveNode,
  isElementEnd,
  isElementNode,
  isEventNode,
  isExpressionNode,
  isFragmentNode,
  isRefNode,
  isSpreadNode,
  isStructuralNode,
  isTextNode,
} from '../ast';
import {
  serializeChildren,
  serializeComponentProp,
  serializeCreateComponent,
  serializeParentExpression,
} from '../jsx/utils';
import type { ASTSerializer } from '../transform';

const ID = {
  template: '$$_templ',
  root: '$$_root',
  walker: '$$_walker',
  element: '$$_el',
  host: '$$_host',
  component: '$$_comp',
  expression: '$$_expr',
};

const RUNTIME = {
  createTemplate: '$$_create_template',
  createFragment: '$$_create_fragment',
  createComponent: '$$_create_component',
  createWalker: '$$_create_walker',
  nextTemplate: '$$_next_template',
  nextElement: '$$_next_element',
  nextCustomElement: '$$_next_custom_element',
  hostElement: '$$_host_element',
  createElement: '$$_create_element',
  setupCustomElement: '$$_setup_custom_element',
  insert: '$$_insert',
  insertAtMarker: '$$_insert_at_marker',
  listen: '$$_listen',
  delegateEvents: '$$_delegate_events',
  clone: '$$_clone',
  directive: '$$_directive',
  ref: '$$_ref',
  attr: '$$_attr',
  class: '$$_class',
  style: '$$_style',
  spread: '$$_spread',
  mergeProps: '$$_merge_props',
  innerHTML: '$$_inner_html',
  computed: '$$_computed',
  effect: '$$_effect',
  peek: '$$_peek',
};

const MARKER = {
  component: '<!$>',
  element: '<!$>',
  expression: '<!$>',
};

const NEXT_ELEMENT = createFunctionCall(RUNTIME.nextElement, [ID.walker]);
const NEXT_MARKER = `${ID.walker}.nextNode()`;

export const dom: ASTSerializer = {
  name: 'dom',
  serialize(ast, ctx) {
    const firstNode = ast.tree[0],
      isFirstNodeElement = isElementNode(firstNode),
      isFirstNodeExpression = isExpressionNode(firstNode),
      isFirstNodeHostElement = isFirstNodeElement && firstNode.tagName === 'HostElement',
      isFirstNodeComponent = isFirstNodeElement && firstNode.isComponent;

    if (
      isFirstNodeExpression &&
      !firstNode.dynamic &&
      !firstNode.children?.length &&
      !firstNode.observable
    ) {
      return firstNode.value;
    }

    let skip = -1,
      initRoot = false,
      innerHTML = false,
      childrenFragment = false,
      currentId!: string,
      component!: ElementNode | undefined,
      definition!: AttributeNode | undefined,
      hostElement!: ElementNode | undefined,
      customElement!: ElementNode | undefined,
      templateId: string | undefined,
      returnId: string | undefined,
      hierarchy: number[] = [],
      props: string[] = [],
      styles: string[] = [],
      spreads: string[] = [],
      expressions: string[] = [],
      template: string[] = [],
      elements: ElementNode[] = [],
      locals = new Declarations(),
      elementChildIndex = -1,
      elementIds: Record<string, string> = {},
      createRootId = () => {
        if (!initRoot) {
          if (!templateId) templateId = ctx.globals.create(ID.template);

          if (ctx.hydratable) {
            locals.create(
              `[${ID.root}, ${ID.walker}]`,
              createFunctionCall(RUNTIME.createWalker, [templateId]),
            );
            ctx.runtime.add(RUNTIME.createWalker);
          } else {
            locals.create(ID.root, createFunctionCall(RUNTIME.clone, [templateId]));
          }

          currentId = ID.root;
          ctx.runtime.add(RUNTIME.clone);
          elementIds[0] = ID.root;
          initRoot = true;
        }

        return ID.root;
      },
      nextElement = () => {
        createRootId();
        ctx.runtime.add(RUNTIME.nextElement);
        return NEXT_ELEMENT;
      },
      nextMarker = () => {
        createRootId();
        return NEXT_MARKER;
      },
      getParentElementId = () => {
        createRootId();
        return getElementId(hierarchy, elementIds, locals);
      },
      getCurrentElementId = () => {
        createRootId();
        return getElementId(
          elementChildIndex >= 0 ? [...hierarchy, elementChildIndex] : hierarchy,
          elementIds,
          locals,
        );
      },
      getNextElementId = () => {
        createRootId();
        const element = elements.at(-1);
        const nextSibling = elementChildIndex + 1;
        return element && element.childCount > 1
          ? nextSibling >= element.childElementCount
            ? 'null'
            : getElementId([...hierarchy, nextSibling], elementIds, locals)
          : null;
      },
      addAttrExpression = (node: AttributeNode, runtimeId: string, name: string = node.name) => {
        if (node.observable) {
          expressions.push(
            createFunctionCall(RUNTIME.effect, [
              `() => ${createFunctionCall(runtimeId, [
                currentId,
                createStringLiteral(name),
                node.value,
              ])}`,
            ]),
          );

          ctx.runtime.add(RUNTIME.effect);
        } else {
          expressions.push(
            createFunctionCall(runtimeId, [currentId, createStringLiteral(name), node.value]),
          );
        }

        ctx.runtime.add(runtimeId);
      },
      addChildren = (children: ComponentChildren[]) => {
        const scoped = children.length > 1 || !isAST(children[0]);
        const serialized = serializeChildren(dom, children, { ...ctx, scoped }, true);
        const hasReturn = scoped || /^(\[|\(|\$\$|\")/.test(serialized);
        props.push(`get $children() { ${hasReturn ? `return ${serialized}` : serialized} }`);
      },
      insert = (marker: (() => string) | null, block: string) => {
        const beforeId = ctx.hydratable ? null : getNextElementId();
        const parentId = ctx.hydratable ? marker?.() ?? null : getParentElementId();
        const insertId = ctx.hydratable ? RUNTIME.insertAtMarker : RUNTIME.insert;
        expressions.push(createFunctionCall(insertId, [parentId, block, beforeId]));
        ctx.runtime.add(insertId);
      },
      isVirtualElement = () => customElement || hostElement;

    if (
      isFirstNodeElement &&
      !firstNode.isComponent &&
      firstNode.tagName !== 'HostElement' &&
      firstNode.tagName !== 'CustomElement'
    ) {
      hierarchy.push(0);
      if (ctx.hydratable) {
        template.push(MARKER.element);
      }
    }

    for (let i = 0; i < ast.tree.length; i++) {
      if (i <= skip) continue;

      const node = ast.tree[i];

      if (component) {
        if (isAttributeNode(node) && !node.namespace) {
          props.push(serializeComponentProp(dom, node, ctx));
        } else if (isSpreadNode(node)) {
          props.push('$$SPREAD');
          spreads.unshift(node.value);
        } else if (isStructuralNode(node) && isElementEnd(node)) {
          if (component.children) addChildren(component.children);

          const { createComponent, shouldMergeProps } = serializeCreateComponent(
            RUNTIME.createComponent,
            RUNTIME.mergeProps,
            component.tagName,
            props,
            spreads,
          );

          if (isFirstNodeComponent) {
            expressions.push(createComponent);
          } else {
            insert(() => currentId, createComponent);
          }

          ctx.runtime.add(RUNTIME.createComponent);
          if (shouldMergeProps) ctx.runtime.add(RUNTIME.mergeProps);

          props = [];
          spreads = [];
          component = undefined;
        }
      } else if (isStructuralNode(node)) {
        if (isAttributesEnd(node)) {
          if (hostElement) continue;

          if (customElement) {
            if (!innerHTML && customElement.children) {
              addChildren(customElement.children);
            }

            expressions.push(
              createFunctionCall(RUNTIME.setupCustomElement, [
                currentId,
                definition!.value,
                props.length > 0 ? `{ ${props.join(', ')} }` : null,
              ]),
            );

            ctx.runtime.add(RUNTIME.setupCustomElement);
          }

          if (styles.length > 0) {
            template.push(` style="${styles.join(';')}"`);
            styles = [];
          }

          const element = elements.at(-1);
          if (element && !element.isVoid && !customElement) template.push('>');
        } else if (isChildrenStart(node)) {
          if (innerHTML) {
            let depth = 0;
            for (let j = i + 1; j < ast.tree.length; j++) {
              const node = ast.tree[j];
              if (isStructuralNode(node)) {
                if (isChildrenStart(node)) {
                  depth++;
                } else if (isChildrenEnd(node)) {
                  if (depth === 0) {
                    skip = j + 1;
                    break;
                  } else {
                    depth--;
                  }
                }
              }
            }

            const element = elements.pop();
            if (element) {
              template.push(element.isVoid ? ` />` : `</${element.tagName}>`);
            }

            elementChildIndex = hierarchy.pop()!;
          } else if (elements.at(-1) !== firstNode) {
            hierarchy.push(elementChildIndex);
            elementChildIndex = -1;
          }
        } else if (isChildrenEnd(node)) {
          elementChildIndex = hierarchy.pop()!;
        } else if (isElementEnd(node)) {
          if (hostElement) {
            if (expressions.length === 0) locals.delete(ID.host);

            if (hostElement.children) {
              expressions.push(
                serializeChildren(dom, hostElement.children, { ...ctx, scoped: true }),
              );
              childrenFragment = true;
            }

            hostElement = undefined;
            continue;
          }

          if (customElement) {
            if (!ctx.hydratable) {
              if (initRoot) {
                insert(null, currentId);
              } else {
                returnId = currentId;
              }
            } else if (elements.length === 0) {
              returnId = currentId;
            }

            definition = undefined;
            customElement = undefined;
          } else {
            const element = elements.pop();
            if (element) {
              template.push(element.isVoid ? ` />` : `</${element.tagName}>`);
            }
          }

          innerHTML = false;
        }
      } else if (isFragmentNode(node)) {
        if (node.children) {
          expressions.push(
            serializeChildren(dom, node.children, {
              ...ctx,
              scoped: true,
              fragment: true,
            }),
          );

          childrenFragment = true;
        } else {
          expressions.push('""');
        }
      } else if (isElementNode(node)) {
        if (node.tagName === 'HostElement') {
          currentId = locals.create(ID.host, createFunctionCall(RUNTIME.hostElement));
          ctx.runtime.add(RUNTIME.hostElement);
          hostElement = node;
          continue;
        }

        if (isFirstNodeComponent && node == firstNode) {
          component = node;
          continue;
        }

        if (node.isComponent) component = node;
        if (node.tagName === 'CustomElement') customElement = node;

        const isElement = !component && !customElement;
        if (i > 0 && isElement) elementChildIndex++;

        const dynamic = node.dynamic();

        if (customElement) {
          if (elements.length > 0) createRootId();

          definition = findCustomElementDefinition(ast, node, i);
          currentId = locals.create(
            ID.element,
            ctx.hydratable
              ? createFunctionCall(RUNTIME.nextCustomElement, [
                  definition.value,
                  elements.length > 0 ? ID.walker : null,
                ])
              : createFunctionCall(RUNTIME.createElement, [`${definition.value}.tagName`]),
          );

          if (ctx.hydratable) {
            ctx.runtime.add(RUNTIME.nextCustomElement);
          } else {
            ctx.runtime.add(RUNTIME.createElement);
          }
        } else if (dynamic) {
          if (ctx.hydratable) {
            if (node.isComponent) {
              currentId = locals.create(ID.component, nextMarker());
            } else if (i > 0) {
              currentId = locals.create(ID.element, nextElement());
            } else {
              currentId = createRootId();
            }
          } else {
            currentId = getCurrentElementId();
          }
        }

        if (ctx.hydratable && i > 0 && dynamic) {
          template.push(MARKER.element);
        }

        if (isElement) {
          template.push(`<${node.tagName}`);
          elements.push(node);
        }
      } else if (isAttributeNode(node)) {
        if (node.namespace) {
          if (node.namespace === '$prop') {
            if (node.name === 'innerHTML') innerHTML = true;
            if (customElement) {
              props.push(serializeComponentProp(dom, node, ctx));
            } else if (node.name === 'innerHTML') {
              expressions.push(createFunctionCall(RUNTIME.innerHTML, [currentId, node.value]));
              ctx.runtime.add(RUNTIME.innerHTML);
            } else if (node.observable) {
              expressions.push(
                createFunctionCall(RUNTIME.effect, [
                  `() => void (${currentId}.${node.name} = ${node.value})`,
                ]),
              );
              ctx.runtime.add(RUNTIME.effect);
            } else {
              expressions.push(`${currentId}.${node.name} = ${node.value};`);
            }
          } else if (node.namespace === '$class') {
            addAttrExpression(node, RUNTIME.class);
          } else if (node.namespace === '$style') {
            if (!node.dynamic && !isVirtualElement()) {
              styles.push(`${node.name}: ${trimQuotes(node.value)}`);
            } else {
              addAttrExpression(node, RUNTIME.style);
            }
          } else if (node.namespace === '$cssvar') {
            if (!node.dynamic && !isVirtualElement()) {
              styles.push(`--${node.name}: ${trimQuotes(node.value)}`);
            } else {
              addAttrExpression(node, RUNTIME.style, `--${node.name}`);
            }
          }
        } else if (!node.dynamic && !isVirtualElement()) {
          if (node.name === 'style') {
            styles.push(trimTrailingSemicolon(trimQuotes(node.value)));
          } else {
            template.push(` ${node.name}="${escape(trimQuotes(node.value), true)}"`);
          }
        } else if (!isVirtualElement() || node.name !== '$element') {
          addAttrExpression(node, RUNTIME.attr);
        }
      } else if (isRefNode(node)) {
        expressions.push(createFunctionCall(RUNTIME.ref, [currentId, node.value]));
        ctx.runtime.add(RUNTIME.ref);
      } else if (isEventNode(node)) {
        if (node.delegate && ctx.delegateEvents) {
          ctx.events.add(createStringLiteral(node.type));
          expressions.push(`${currentId}.$$${node.type} = ${node.value};`);
          if (node.data) {
            expressions.push(`${currentId}.$$${node.type}Data = ${node.data};`);
          }
        } else {
          const args = [currentId, createStringLiteral(node.type), node.value];
          if (node.namespace === '$oncapture') args.push(`1 /* CAPTURE */`);
          expressions.push(createFunctionCall(RUNTIME.listen, args));
          ctx.runtime.add(RUNTIME.listen);
        }
      } else if (isDirectiveNode(node)) {
        expressions.push(createFunctionCall(RUNTIME.directive, [currentId, node.name, node.value]));
        ctx.runtime.add(RUNTIME.directive);
      } else if (isTextNode(node)) {
        if (!ctx.hydratable) elementChildIndex++;
        template.push(node.value);
      } else if (isExpressionNode(node)) {
        if (!node.dynamic && !isVirtualElement()) {
          template.push(encode(trimQuotes(node.value)));
        } else {
          const shouldInsert = !isFirstNodeExpression || node !== firstNode;
          if (!initRoot && shouldInsert) createRootId();
          if (ctx.hydratable && shouldInsert) template.push(MARKER.expression);
          const code = !node.children
            ? node.value
            : serializeParentExpression(dom, node, ctx, !shouldInsert && RUNTIME.peek);
          const expression = node.callId ?? (node.observable ? `() => ${code}` : code);
          if (shouldInsert) {
            insert(() => locals.create(ID.expression, nextMarker()), expression);
          } else if (ctx.fragment && ctx.hydratable) {
            if (!node.observable && !node.children?.length) {
              expressions.push(expression);
            } else if (node.children?.length) {
              expressions.push(
                selfInvokingFunction(
                  [
                    `const $$_signal = ${createFunctionCall(RUNTIME.computed, [`() => ${code}`])};`,
                    '$$_signal();',
                    'return $$_signal;',
                  ].join(''),
                ),
              );
              ctx.runtime.add(RUNTIME.computed);
            } else {
              expressions.push(
                node.callId ?? createFunctionCall(RUNTIME.computed, [`() => ${code}`]),
              );
              ctx.runtime.add(RUNTIME.computed);
            }
          } else {
            expressions.push(expression);
            ctx.runtime.add(RUNTIME.peek);
          }
        }
      } else if (isSpreadNode(node)) {
        expressions.push(createFunctionCall(RUNTIME.spread, [currentId, node.value]));
        ctx.runtime.add(RUNTIME.spread);
      }
    }

    if (template.length) {
      const expression = createFunctionCall(RUNTIME.createTemplate, [`\`${template.join('')}\``]);

      if (templateId) {
        ctx.globals.update(templateId, expression);
      } else {
        templateId = ctx.globals.create(ID.template, expression);
      }

      ctx.runtime.add(RUNTIME.createTemplate);
    } else if (templateId) {
      ctx.globals.update(templateId, createFunctionCall(RUNTIME.createFragment));
      ctx.runtime.add(RUNTIME.createFragment);
    }

    if (locals.size > 1 || expressions.length) {
      return isFirstNodeComponent || isFirstNodeExpression
        ? expressions.join(';')
        : childrenFragment && expressions.length === 1
        ? expressions[expressions.length - 1]
        : [
            ctx.scoped && `(() => { `,
            locals.serialize(),
            '\n',
            ...(childrenFragment ? expressions.slice(0, -1) : expressions).join(';'),
            ';',
            '\n',
            '\n',
            `return ${
              isFirstNodeHostElement
                ? childrenFragment
                  ? expressions[expressions.length - 1]
                  : ID.host
                : returnId ?? createRootId()
            }`,
            ctx.scoped && '})()',
          ]
            .filter(Boolean)
            .join('');
    } else if (templateId) {
      if (ctx.hydratable) {
        ctx.runtime.add(RUNTIME.nextTemplate);
        return createFunctionCall(RUNTIME.nextTemplate, [templateId]);
      }

      ctx.runtime.add(RUNTIME.clone);
      return createFunctionCall(RUNTIME.clone, [templateId]);
    }

    return 'null';
  },
};

function getElementId(
  positions: number[],
  cache: Record<string, string>,
  declarations: Declarations,
) {
  const key = positions.join('');
  if (cache[key]) return cache[key];

  let id = ID.root,
    hierarchy = '0';

  for (let i = 1; i < positions.length; i++) {
    const childIndex = positions[i];
    const current = hierarchy + childIndex;

    if (cache[current]) {
      id = cache[current];
    } else {
      for (let j = 0; j <= childIndex; j++) {
        const sibling = hierarchy + j;
        id = cache[sibling] ??= declarations.create(
          ID.element,
          `${id}.${j === 0 ? 'firstChild' : 'nextSibling'}`,
        );
      }
    }

    hierarchy = current;
  }

  return id;
}

function findCustomElementDefinition(ast: AST, node: ElementNode, start: number) {
  for (let j = start; j < ast.tree.length; j++) {
    const attr = ast.tree[j];
    if (isAttributeNode(attr) && attr.name === '$element') {
      return attr;
    } else if (isStructuralNode(attr)) {
      break;
    }
  }

  const loc = kleur.bold(
    `${node.ref.getSourceFile().fileName} ${kleur.cyan(
      `${node.ref.getStart()}:${node.ref.getEnd()}`,
    )}`,
  );

  throw Error(`[maverick] \`element\` prop was not provided for \`CustomElement\` at ${loc}`);
}
