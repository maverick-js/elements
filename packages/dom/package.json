{
  "name": "@maverick-js/dom",
  "license": "MIT",
  "version": "0.43.2",
  "type": "module",
  "types": "src/index.ts",
  "module": "src/index.ts",
  "sideEffects": false,
  "contributors": [
    "Rahim Alwer <rahim.alwer@gmail.com>"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/maverick-js/maverick.git",
    "directory": "packages/dom"
  },
  "bugs": {
    "url": "https://github.com/maverick-js/maverick/issues"
  },
  "scripts": {
    "dev": "rollup -c -w",
    "build": "pnpm clean && pnpm types && rollup -c",
    "clean": "rimraf dist-npm",
    "types": "tsc -p tsconfig.build.json",
    "format": "prettier src --write --loglevel warn",
    "test:pre": "node .scripts/test-pre.js",
    "test": "pnpm test:pre && vitest -c vite.config.js --run",
    "test:watch": "pnpm test:pre && vitest -c vite.config.js --watch"
  },
  "dependencies": {
    "@maverick-js/std": "workspace:*"
  },
  "devDependencies": {
    "@edge-runtime/vm": "^2.1.2",
    "@maverick-js/compiler": "workspace:*",
    "@maverick-js/core": "workspace:*",
    "@maverick-js/element": "workspace:*",
    "@rollup/plugin-node-resolve": "^15.3.0",
    "@vitest/browser": "^2.1.3",
    "esbuild": "^0.19.12",
    "playwright": "^1.48.1",
    "rimraf": "^5.0.10",
    "rollup": "^4.24.0",
    "rollup-plugin-dts": "^6.1.1",
    "rollup-plugin-esbuild": "^6.1.1",
    "type-fest": "^4.26.1",
    "typescript": "^5.6.3",
    "vite": "^5.4.9",
    "vitest": "^2.1.3"
  },
  "peerDependencies": {
    "@maverick-js/core": "workspace:*"
  },
  "exports": {
    ".": "./src/index.ts"
  },
  "$types": "index.d.ts",
  "$exports": {
    ".": {
      "types": "./index.d.ts",
      "node": "./server/index.js",
      "development": "./dev/index.js",
      "default": "./prod/index.js"
    },
    "./server": {
      "types": "./server.d.ts",
      "node": "./server/server.js",
      "development": "./dev/server.js",
      "default": "./prod/server.js"
    },
    "./types/*": "./types/*",
    "./package.json": "./package.json"
  },
  "publishConfig": {
    "access": "public"
  }
}
