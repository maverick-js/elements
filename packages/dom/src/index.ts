export * from './insert';
export * from './reconcile';
export * from './render';
export * from './runtime';
export * from './walker';

export * from './components/for';
export * from './components/fragment';
export * from './components/host';
export * from './components/portal';
