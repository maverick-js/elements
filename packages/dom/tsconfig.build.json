{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "declarationDir": "types",
    "paths": {
      "@maverick-js/compiler": ["../compiler/types/index.d.ts"],
      "@maverick-js/std": ["../std/types/index.d.ts"],
      "maverick.js": ["../maverick/types/core/index.d.ts"]
    }
  },
  "include": ["src"]
}
