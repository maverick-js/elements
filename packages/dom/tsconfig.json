{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "lib": ["dom", "dom.iterable"],
    "types": ["vitest/globals"]
  },
  "include": ["src", "tests"],
  "exclude": ["dist-*", "node_modules", "*.d.ts"]
}
