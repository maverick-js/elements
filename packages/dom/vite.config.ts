/// <reference types="vitest" />
import { defineConfig } from 'vite';

const SERVER = !!process.env.SERVER;

export default defineConfig({
  define: {
    __DEV__: 'true',
    __TEST__: 'true',
    __SERVER__: SERVER ? 'true' : 'false',
  },
  resolve: {
    alias: {
      '@maverick-js/dom': '/src/index.ts',
    },
  },
  // https://vitest.dev/config
  test: {
    include: [`tests/${SERVER ? 'server' : 'client'}/**/*.test.{ts,tsx}`],
    globals: true,
    environment: SERVER ? 'edge-runtime' : 'jsdom',
  },
});
