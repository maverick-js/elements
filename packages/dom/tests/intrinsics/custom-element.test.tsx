import { render } from '@maverick-js/dom';
import { type CustomElementOptions, Host, MaverickComponent } from 'maverick.js';

const target = document.body;

afterEach(() => {
  target.textContent = '';
});

class Foo extends MaverickComponent {
  static element: CustomElementOptions = {
    name: 'mk-foo',
  };

  override render() {
    return (
      <Host data-foo as="div">
        <div>Foo Content</div>
        <Bar />
      </Host>
    );
  }
}

class Bar extends MaverickComponent {
  static element: CustomElementOptions = {
    name: 'mk-bar',
  };

  override render() {
    return (
      <Host data-bar as="div">
        <div>Bar Content</div>
      </Host>
    );
  }
}

test('render', () => {
  render(() => <Foo />, { target });
  expect(target).toMatchSnapshot();
});
