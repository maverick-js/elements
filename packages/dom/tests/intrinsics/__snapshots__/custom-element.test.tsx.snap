// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render 1`] = `
<body>
  
    
  
  


  <mk-foo
    data-foo=""
    data-maverick=""
  >
    <div>
      Foo Content
    </div>
    <mk-bar
      data-bar=""
      data-maverick=""
    >
      <div>
        Bar Content
      </div>
    </mk-bar>
    <mk-hux
      data-hux=""
      data-maverick=""
    />
  </mk-foo>
</body>
`;

exports[`render 2`] = `
HTMLCollection [
  <div>
    Hux Content
  </div>,
]
`;
